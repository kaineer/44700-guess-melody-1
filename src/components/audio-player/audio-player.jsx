import React, {Component, createRef} from 'react';
import {string, bool, func} from 'prop-types';

export const snapshotDummyURL = `void://for-snapshot`;

export class AudioPlayer extends Component {
  constructor(props) {
    super(props);

    this._audioRef = createRef();

    this.state = {
      progress: 0,
      isLoading: false,
    };

    this._onPlayButtonClick = this._onPlayButtonClick.bind(this);
  }

  componentDidMount() {
    if (this.props.src !== snapshotDummyURL) {
      this._setAudio();
    }
  }

  componentWillUnmount() {
    const audio = this._audio;

    Object.assign(audio, {
      oncanplaythrough: null,
      onplay: null,
      onpause: null,
      ontimeupdate: null,
      src: null,
    });

    this._audio = null;
  }

  componentDidUpdate() {
    const {state: {isLoading}, props: {isPlaying}} = this;
    const audio = this._audio;

    if (!isLoading) {
      if (isPlaying) {
        audio.play();
      } else {
        audio.pause();
      }
    }
  }

  render() {
    const {state: {isLoading}, props: {isPlaying}} = this;
    const buttonModifier = `track__button--${isPlaying ? `pause` : `play`}`;

    return (
      <div className="game__track">
        <button
          className={`track__button ${buttonModifier}`}
          type="button"
          disabled={isLoading}
          onClick={this._onPlayButtonClick}
        />
        <div className="track__status">
          <audio ref={this._audioRef} />
        </div>
      </div>
    );
  }

  _setAudio() {
    const {props: {src, isPlaying}} = this;

    this._audio = new Audio(src);

    const audio = this._audio;

    const setState = this.setState.bind(this);

    setState({
      progress: audio.currentTime,
      isLoading: true
    });

    Object.assign(audio, {
      oncanplaythrough() {
        setState({isLoading: false, isPlaying});
      },
      onplay() {
        setState({isPlaying: true});
      },
      onpause() {
        setState({isPlaying: false});
      },
      ontimeupdate() {
        setState({progress: audio.currentTime});
      },
      onended() {
        setState({progress: 0, isPlaying: false});
      }
    });
  }

  _onPlayButtonClick() {
    const {isPlaying} = this.props;
    this.props.onTogglePlaying(!isPlaying);
  }
}

AudioPlayer.propTypes = {
  isPlaying: bool.isRequired,
  onTogglePlaying: func.isRequired,
  src: string.isRequired,
};
