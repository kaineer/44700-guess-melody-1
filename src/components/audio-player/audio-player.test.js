import React from 'react';
import {AudioPlayer, snapshotDummyURL} from './audio-player';
import {create} from 'react-test-renderer';


describe(`<AudioPlayer />`, () => {
  it(`should render correctly`, () => {
    const tree = create(
        <AudioPlayer
          src={snapshotDummyURL}
          isPlaying={false}
          onTogglePlaying={jest.fn()}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
