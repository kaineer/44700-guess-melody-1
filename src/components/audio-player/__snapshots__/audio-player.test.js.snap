// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AudioPlayer /> should render correctly 1`] = `
<div
  className="game__track"
>
  <button
    className="track__button track__button--play"
    disabled={false}
    onClick={[Function]}
    type="button"
  />
  <div
    className="track__status"
  >
    <audio />
  </div>
</div>
`;
